README.md
# wsc-api-examples-ruby

## Wowza Streaming Cloud REST API - Example Application for the Live Stream Workflow Written in Ruby

This API example application demonstrates the live stream workflow using the `/live_streams` endpoint of the Wowza Streaming Cloud™ service REST API version 1.

> **Note:** Wowza Streaming Cloud REST API version 1.0 is in a public preview release and isn't intended for use in production environments. All use of the Wowza Streaming Cloud REST API is subject to the Wowza Streaming Cloud Terms of Use.

If you just want to know how to use the Wowza Streaming Cloud API using the Ruby Net::HTTP library, [look at this section](#how_to_ruby_net_http).

## Contents
- [Prerequisites](#prerequisites)
- [Installation](#installation)
- [Configuration](#configuration)
- [Run the application](#run)
- [How to use the example application](#how_to_use)
- [How to use the Wowza Streaming Cloud REST API with the Ruby Net::HTTP library](#how_to_ruby_net_http)
- [References](#references)
- [Contact](#contact)
- [License](#license)

<a name="prerequisites"></a>
## Prerequisites

To install the example application make sure you have [Ruby](https://www.ruby-lang.org/en/documentation/installation/), [RubyGems](https://rubygems.org/), and [Bundler](http://bundler.io/) ready to use on your system.

<a name="installation"></a>
## Installation

1. In Terminal, make sure you're in your application directory:

```bash
$ cd /wsc-api-example-ruby
```

2. Run bundle install:

```bash
$ bundle install
```

You're done!

<a name="configuration"></a>
## Configuration

To use the application, you need a valid [Wowza Streaming Cloud account](https://cloud.wowza.com/) and access to the API.

> **Note:** If you don't have API access, [Request Access to Wowza Streaming Cloud API](https://www.wowza.com/products/streaming-cloud/features/api-access-request). After you're accepted to the Wowza Streaming Cloud REST API public preview, your API key will be available on the **API Access** page of your [Wowza Streaming Cloud account](https://cloud.wowza.com/).

### Put the 'API Key' and the 'API Access Key' into the configuration file

```yml
# config/keys.yml
api_key: "- PASTE YOUR API KEY HERE -"
api_access_key: "- PASTE YOUR API ACCESS KEY HERE -"
```

### (Optional) Change the hostname of the Wowza Streaming Cloud environment

You have two options. To use our sandbox environment and be safe, use this hostname:

```yml
# config/settings.yml
api_base_url: "https://api-sandbox.cloud.wowza.com"
...
```

> **Note:** This is the default. If you don't change anything, you'll use the free sandbox environment.

If you know what to do and want to use your live account (**and accrue charges**), use this hostname:

```yml
# config/settings.yml
api_base_url: "https://api.cloud.wowza.com"
...
```

### (Optional) Enable debug output

If you want to see header response data and response codes for each API call, enable it in the settings:

```yml
# config/settings.yml
...
debug: true
```

This is disabled by default.

<a name="run"></a>
## Run the application

In Terminal, make sure you're in your application directory and execute the Ruby file:

```bash
$ ./live_stream_api_example.rb
```

> **Note:** Make sure the file is executable: ```$ chmod +x live_stream_api_example.rb```

That's all.

<a name="how_to_use"></a>
## How to use the example application

After launching the application, you'll see the main menu with several action items. Just enter a number and press Return to be guided through the application. It's easy.

*The screen after launching the application:*
```
"Main Menu"

1. Show the number of live streams in your account
2. List all live streams of your account
3. Create a live stream with pre-configured settings         => data/live_stream/encoder_types/*.json
4. Show the details of an existing live stream
5. Update a live stream with pre-configured settings         => data/live_stream/update_example.json
6. Start a live stream                                       => only for live streams with the 'stopped' state
7. Reset a live stream                                       => only for live streams with the 'started' state
8. Stop a live stream                                        => only for live streams with the 'started' state
9. Show the current state of a live stream
10. Show the thumbnail URL of a live stream                  => only for live streams with the 'started' state
11. Delete a live stream
12. Run the pre-configured live stream workflow
13. Quit :(

Enter the number for the action you want to execute.
```

**Actions 1 through 11** use single API calls to your account's data. This should give you an overview what you can do with the `/live_streams` endpoint on Wowza Streaming Cloud.

**Action 12** ('Run the pre-configured live stream workflow') is meant to give you an idea of the complete live stream workflow. It uses the same code as actions 1-11. It's separated into the following steps:

1. Create a live stream
2. Start the live stream
3. Poll the status of the live stream (to know when the live stream in your requested location is ready to use)
4. Poll the status of the player that was created with your live stream (to know when the player and the hosted page are successfully provisioned and ready to use)
5. Receive the hosted page URL (where you can watch the stream)
6. Stop the live stream
7. Delete the live stream

You will be guided through each step.

> **Note:** The live stream settings used by default in this example application are based on the **Other RTSP Pull** encoder setting described in our user interface on [https://cloud.wowza.com/](https://cloud.wowza.com/). The data, along with lots of other examples of different live stream configurations, are stored in JSON files in the [data/live_stream/](https://github.com/WowzaMediaSystems/wsc-api-examples-ruby/tree/master/data/live_stream) directory in this repository:

```json
# data/live_stream/encoder_types/other_rtsp_pull.json
{
  "live_stream": {
    "name": "My Awesome live stream",
    "transcoder_type": "transcoded",
    "billing_mode": "pay_as_you_go",
    "broadcast_location": "us_west_california",
    "encoder": "other_rtsp",
    "delivery_method": "pull",
    "source_url": "1.2.3.4/axis-media/media.amp",
    "aspect_ratio_width": 1920,
    "aspect_ratio_height": 1080
  }
}
```

<a name="how_to_ruby_net_http"></a>
## How to use the Wowza Streaming Cloud REST API with the Ruby Net::HTTP library

### Get a list of all live streams of your account (GET)

```ruby
require 'net/http'

# build the URI to access the /live_streams endpoint
uri = URI("https://api.cloud.wowza.com/api/v1/live_streams/")

# initialize Net::HTTP using SSL
Net::HTTP.start(uri.host, uri.port, use_ssl: true) do |http|

  # build the HTTP request
  request = Net::HTTP::Get.new uri.path
  # add authentication keys to the request header
  request['wsc-api-key'] = "API_KEY"
  request['wsc-access-key'] = "API_ACCESS_KEY"
  # add the right Content-Type to the header (always required when pushing data)
  request["Content-Type"] = "application/json"

  response = http.request(request)

  # get the response as JSON object
  puts JSON.parse(response.body)
  # get the response code
  puts response.code
  # loop through response headers:
  response.each_header do |key, value|
    puts "#{key} => #{value}"
  end

end

# Equivalent cURL command:
#
# curl -H 'wsc-api-key: KEY' -H 'wsc-access-key: KEY' https://api.cloud.wowza.com/api/v1/live_streams/
```

Example Response:
```json
# 200 OK
{
    "live_streams": [
        {
              "id": "kdgbsvth",
            "name": "My Awesome Live Stream",
             "...": "..."
        },
        {
              "id": "ftklwjvz",
            "name": "My Second Live Stream",
             "...": "..."
        }
    ]
}
```

<a name="how_to_ruby_net_http_more_examples"></a>
## Additional examples

> **Note:** The following sections contain examples that show only the differences from the preceding example.

### Create a live stream (POST)

> **Note:** Check [/data/live_stream/encoder_types/*.json](https://github.com/WowzaMediaSystems/wsc-api-examples-ruby/tree/master/data/live_stream/encoder_types) for real-world JSON examples.

```ruby
uri = URI("https://api.cloud.wowza.com/api/v1/live_streams/")
...
  request = Net::HTTP::Post.new uri.path
  # add the body to the request and parse the hash to JSON (or load it from a JSON file)
  request.body = {
    "live_stream": {
      LIVE_STREAM_SETTINGS
    }
  }.to_json
...

# Equivalent cURL command:
#
# curl -H 'wsc-api-key: KEY' -H 'wsc-access-key: KEY' -H 'Content-Type: application/json' -X POST -d '{
#   "live_stream": {
#     LIVE_STREAM_SETTINGS
#   }
# }' https://api.cloud.wowza.com/api/v1/live_streams/
```

Example Response:
```json
# 201 CREATED
{
    "live_stream": {
          "id": "kdgbsvth",
        "name": "My Awesome Live Stream",
         "...": "..."
    }
}
```

### Details of a live stream (GET)

```ruby
uri = URI("https://api.cloud.wowza.com/api/v1/live_streams/[LIVE_STREAM_ID]/")
...
  request = Net::HTTP::Get.new uri.path
...

# Equivalent cURL command:
#
# curl -H 'wsc-api-key: KEY' -H 'wsc-access-key: KEY' https://api.cloud.wowza.com/api/v1/live_streams/[LIVE_STREAM_ID]/
```

Example Response:
```json
# 200 OK
{
    "live_stream": {
                                   "id": "kdgbsvth",
                                 "name": "My Awesome Live Stream",
                      "transcoder_type": "transcoded",
                         "billing_mode": "pay_as_you_go",
                   "broadcast_location": "us_west_california",
                            "recording": false,
                  "closed_caption_type": "none",
                              "encoder": "other_rtsp",
                      "delivery_method": "pull",
                    "use_stream_source": false,
                   "aspect_ratio_width": 1920,
                  "aspect_ratio_height": 1080,
        "source_connection_information": { "source_url": "rtsp://..." },
                            "player_id": "g7kpkj1n",
                    "player_responsive": false,
                         "player_width": 640,
                     "player_countdown": false,
                    "player_embed_code": "in_progress",
              "player_hds_playback_url": "http://...",
              "player_hls_playback_url": "http://...",
                          "hosted_page": true,
                    "hosted_page_title": "My Awesome Event",
                      "hosted_page_url": "https://...",
            "hosted_page_sharing_icons": true,
                           "created_at": "2016-02-29T17:59:25.251Z",
                           "updated_at": "2016-02-29T17:59:25.604Z"
    }
}
```

### Update a live stream (PATCH)

> **Note:** Check [/data/live_stream/update_example.json](https://github.com/WowzaMediaSystems/wsc-api-examples-ruby/tree/master/data/live_stream/update_example.json) for a JSON example.

```ruby
uri = URI("https://api.cloud.wowza.com/api/v1/live_streams/[LIVE_STREAM_ID]/")
...
  request = Net::HTTP::Patch.new uri.path
  # add the body to the request and parse the hash to JSON (or load it from a JSON file)
  request.body = {
    "live_stream": {
      LIVE_STREAM_SETTINGS
    }
  }.to_json
...

# Equivalent cURL command:
#
# curl -H 'wsc-api-key: KEY' -H 'wsc-access-key: KEY' -H 'Content-Type: application/json' -X PATCH -d '{
#   "live_stream": {
#     LIVE_STREAM_SETTINGS
#   }
# }' https://api.cloud.wowza.com/api/v1/live_streams/[LIVE_STREAM_ID]/
```

Example Response:
```json
# 200 OK
{
    "live_stream": {
                                   "id": "kdgbsvth",
                                 "name": "My Updated Awesome Live Stream",
                                  "...": "..."
    }
}
```

### Start a live stream (PUT)

```ruby
uri = URI("https://api.cloud.wowza.com/api/v1/live_streams/[LIVE_STREAM_ID]/start/")
...
  request = Net::HTTP::Put.new uri.path
...

# Equivalent cURL command:
#
# curl -H 'wsc-api-key: KEY' -H 'wsc-access-key: KEY' -X PUT -d '' https://api.cloud.wowza.com/api/v1/live_streams/[LIVE_STREAM_ID]/start/
```

Example Response:
```json
# 200 OK
{
    "live_stream": {
        "state": "starting"
    }
}
```

### Reset a live stream (PUT)

```ruby
uri = URI("https://api.cloud.wowza.com/api/v1/live_streams/[LIVE_STREAM_ID]/reset/")
...
  request = Net::HTTP::Put.new uri.path
...

# Equivalent cURL command:
#
# curl -H 'wsc-api-key: KEY' -H 'wsc-access-key: KEY' -X PUT -d '' https://api.cloud.wowza.com/api/v1/live_streams/[LIVE_STREAM_ID]/reset/
```

Example Response:
```json
# 200 OK
{
    "live_stream": {
        "state": "resetting"
    }
}
```

### Stop a live stream (PUT)

```ruby
uri = URI("https://api.cloud.wowza.com/api/v1/live_streams/[LIVE_STREAM_ID]/stop/")
...
  request = Net::HTTP::Put.new uri.path
...

# Equivalent cURL command:
#
# curl -H 'wsc-api-key: KEY' -H 'wsc-access-key: KEY' -X PUT -d '' https://api.cloud.wowza.com/api/v1/live_streams/[LIVE_STREAM_ID]/stop/
```

Example Response:
```json
# 200 OK
{
    "live_stream": {
        "state": "stopped"
    }
}
```

### Poll the state of a live stream (GET)

```ruby
uri = URI("https://api.cloud.wowza.com/api/v1/live_streams/[LIVE_STREAM_ID]/state/")
...
  request = Net::HTTP::Get.new uri.path
...

# Equivalent cURL command:
#
# curl -H 'wsc-api-key: KEY' -H 'wsc-access-key: KEY' https://api.cloud.wowza.com/api/v1/live_streams/[LIVE_STREAM_ID]/state/
```

Example Response:
```json
# 200 OK
{
    "live_stream": {
        "state": "started"
    }
}
```

### Poll the state of a player that was created with a live stream (GET)

> **Note:** The player ID can be found in the property 'player_id' in a live stream JSON response.

```ruby
uri = URI("https://api.cloud.wowza.com/api/v1/players/[PLAYER_ID]/state/")
...
  request = Net::HTTP::Get.new uri.path
...

# Equivalent cURL command:
#
# curl -H 'wsc-api-key: KEY' -H 'wsc-access-key: KEY' https://api.cloud.wowza.com/api/v1/players/[PLAYER_ID]/state/
```

Example Response:
```json
# 200 OK
{
    "player": {
        "state": "activated"
    }
}
```

### Show thumbnail URL (preview image) of a live stream (GET)

```ruby
uri = URI("https://api.cloud.wowza.com/api/v1/live_streams/[LIVE_STREAM_ID]/thumbnail_url/")
...
  request = Net::HTTP::Get.new uri.path
...

# Equivalent cURL command:
#
# curl -H 'wsc-api-key: KEY' -H 'wsc-access-key: KEY' https://api.cloud.wowza.com/api/v1/live_streams/[LIVE_STREAM_ID]/thumbnail_url/
```

Example Response:
```json
# 200 OK
{
    "live_stream": {
        "thumbnail_url": "http://..."
    }
}
```

### Delete a live stream (DELETE)

```ruby
uri = URI("https://api.cloud.wowza.com/api/v1/live_streams/[LIVE_STREAM_ID]/")
...
  request = Net::HTTP::Delete.new uri.path
...

# Equivalent cURL command:
#
# curl -H 'wsc-api-key: KEY' -H 'wsc-access-key: KEY' -X DELETE https://api.cloud.wowza.com/api/v1/live_streams/[LIVE_STREAM_ID]/
```

Example Response:
```json
# 204 NO CONTENT
```

<a name="references"></a>
## References

[Get access to the Wowza Streaming Cloud REST API public preview](https://www.wowza.com/products/streaming-cloud/features/api-access-request)

[Your API keys (API access required)](https://cloud.wowza.com/en/manage/access_keys)

[Interactive API documentation (API access required)](https://sandbox.cloud.wowza.com/apidocs/v1/)

[Wowza Streaming Cloud Support articles](https://www.wowza.com/forums/content.php?775-Wowza-Streaming-Cloud-REST-API)

<a name="contact"></a>
## Contact

Wowza Media Systems™, LLC

Wowza Media Systems provides developers with a platform to create streaming applications and solutions. See [Wowza Developer Tools](https://www.wowza.com/resources/developers) to learn more about our APIs and SDK.

<a name="license"></a>
## License

This is distributed under the [Wowza Public License](https://github.com/WowzaMediaSystems/wsc-api-examples-ruby/blob/master/LICENSE.txt).


Changelog.txt
Date        Description
==========  ====================================================================
2016-03-01  First implementation


live_stream_api_example.rb
#!/usr/bin/env ruby

# This code and all components (c) Copyright 2006 - 2016, Wowza Media Systems, LLC.  All rights reserved.  This code is licensed pursuant to the Wowza Public License version 1.0, available at www.wowza.com/legal.

require 'yaml'
require 'json'
require 'highline'
require 'net/http'
require 'awesome_print'
require 'shell-spinner'

class LiveStreamApiExample

  # init application
  #
  def initialize
    init_config
  end

  # returns the amount of Live Streams in your account, using:
  #
  # GET /live_streams
  #
  # returns the full API response object
  #
  def count
    response = call_api method: :get
    json = JSON.parse(response.body)
    if response.code.to_i==200
      ap("Found #{json['live_streams'].size} Live Streams total.", color: {string: :green})
    else
      ap(JSON.parse(response.body))
    end
    return response
  end

  # returns a list of Live Streams in your account, using:
  #
  # GET /live_streams
  #
  # options:
  #  simple: true|false (optional, default: false)
  #          if set to true a simple list with name and id is returned
  #          if set to false (or nil) the original JSON list is returned
  #
  def list options={}
    response = call_api method: :get
    if options[:simple]
      json = JSON.parse(response.body)
      json['live_streams'].collect{ |ls| puts "#{ls['id']}: #{ls['name']}" }
    else
      ap JSON.parse(response.body), index: false
    end
    return response
  end

  # creates a pre-configured Live Stream, using:
  #
  # POST /live_streams
  #
  # the JSON data with the settings of the Live Stream that is pushed to the API
  # is stored in a file 'data/live_stream.json'
  #
  # returns the full API response object
  #
  def create create_json
    response = call_api method: :post, body: create_json
    ap JSON.parse(response.body)
    return response
  end

  # shows the details of a Live Stream, using:
  #
  # GET /live_streams/[ID]
  #
  # returns the full API response object
  #
  def show uid
    response = call_api method: :get, id: uid
    ap JSON.parse(response.body)
    return response
  end

  # updates a Live Stream with pre-configured settings, using:
  #
  # PATCH /live_streams/[ID]
  #
  # the JSON data with the settings of the Live Stream that is pushed to the API
  # is stored in a file 'data/live_stream/update_example.json'
  #
  # returns the full API response object
  #
  def update uid
    update_json = JSON.parse(File.read('./data/live_stream/update_example.json'))
    response = call_api method: :patch, id: uid, body: update_json
    ap JSON.parse(response.body)
    return response
  end

  # starts a Live Stream, using:
  #
  # PUT /live_streams/[ID]/start
  #
  # returns the full API response object
  #
  def start uid
    response = call_api method: :put, id: uid, action: '/start'
    ap JSON.parse(response.body)
    return response
  end

  # resets a Live Stream, using:
  #
  # PUT /live_streams/[ID]/reset
  #
  # returns the full API response object
  #
  def reset uid
    response = call_api method: :put, id: uid, action: '/reset'
    ap JSON.parse(response.body)
    return response
  end

  # stops a Live Stream, using:
  #
  # PUT /live_streams/[ID]/stop
  #
  # returns the full API response object
  #
  def stop uid
    response = call_api method: :put, id: uid, action: '/stop'
    ap JSON.parse(response.body)
    return response
  end

  # shows the state of a Live Stream, using:
  #
  # GET /live_streams/[ID]/state
  #
  # returns the full API response object
  #
  def state uid
    response = call_api method: :get, id: uid, action: '/state'
    ap JSON.parse(response.body)
    return response
  end

  # shows the thumbnail URL (preview image) of a Live Stream, using:
  #
  # GET /live_streams/[ID]/thumbnail_url
  #
  # returns the full API response object
  #
  def thumbnail_url uid
    response = call_api method: :get, id: uid, action: '/thumbnail_url'
    ap JSON.parse(response.body)
    return response
  end

  # deletes a Live Stream, using:
  #
  # DELETE /live_streams/[ID]
  #
  # returns the full API response object
  #
  def delete uid
    response = call_api method: :delete, id: uid
    ap "#{uid} deleted!", color: {string: :green} if response.code=="204"
    return response
  end

  # shows the state of a Player, using:
  #
  # GET /players/[ID]/state
  #
  # returns the full API response object
  #
  def player_state uid
    response = call_api method: :get, endpoint: 'players', id: uid, action: '/state'
    ap JSON.parse(response.body)
    return response
  end

  # demonstrates a pre-defined workflow of a Live Stream in multiple steps:
  #
  # - create Live Stream
  # - start Live Stream
  # - poll status of Live Stream
  # - poll status of Player
  # - receive hosted page URL of Live Stream
  # - stop Live Stream
  # - delete Live Stream
  #
  def workflow
    next_step_text = "Are you ready? Hit Enter to continue with the next step!"
    next_step = 1

    # First Step: create
    ap "Step #{next_step}: We are going to create a new pre-configured Live Stream.", color: {string: :purpleish}
    puts "\n"
    @cli.ask next_step_text
    # we are going to use the pre-configured 'Other RTSP Pull' example
    encoder_json = JSON.parse(File.read('./data/live_stream/encoder_types/other_rtsp_pull.json'))
    response = create encoder_json
    response_json = JSON.parse(response.body)

    # error handling
    unless response.code=="201"
      puts "\n"
      ap "An error occured. Please check the error above! Returning to Main Menu...", color: {string: :red}
      menu
      return
    end

    # save the id and the player id of the live stream for later use
    id = response_json['live_stream']['id']
    player_id = response_json['live_stream']['player_id']
    next_step += 1

    # Next Step: start
    puts "\n"
    ap "Step #{next_step}: Next we are going to start this Live Stream.", color: {string: :purpleish}
    puts "\n"
    @cli.ask next_step_text
    response = start id
    state = JSON.parse(response.body)['live_stream']['state']
    next_step += 1

    # Next Step: state
    puts "\n"
    ap "Step #{next_step}: Now we are going to check the state of this Live Stream every 10 seconds and wait until it is started.", color: {string: :purpleish}
    puts "\n"
    @cli.ask next_step_text
    while state=='starting' do
      response = state id
      state = JSON.parse(response.body)['live_stream']['state']
      if state=='starting'
        ap "State: '#{state}'. Let's wait a bit and poll again...", color: {string: :yellowish}
        puts "\n"
        sleep 10
      end
    end
    next_step += 1

    # Continue with workflow if Live Stream was started successfully
    if state=="started"
      ap "Nice! Stream #{state}.", color: {string: :green}

      # Next Step: player state
      puts "\n"
      ap "Step #{next_step}: Now we are going to check the state of the player (that was created with the Live Stream) every 10 seconds and wait until the provisioning is finished.", color: {string: :purpleish}
      puts "\n"
      @cli.ask next_step_text
      p_state ='requested'
      while p_state=='requested' do
        response = player_state player_id
        p_state = JSON.parse(response.body)['player']['state']
        if p_state=='requested'
          ap "State: '#{p_state}'. Let's wait a bit and poll again...", color: {string: :yellowish}
          puts "\n"
          sleep 10
        end
      end
      next_step += 1

      # Continue with workflow if Player was provisioned successfully
      if p_state=='activated'
        # Next Step: show
        puts "\n"
        ap "Step #{next_step}: Now we want to see the updated details of the started Live Stream to get the URL of the hosted page.", color: {string: :purpleish}
        puts "\n"
        @cli.ask next_step_text
        response = show id
        hosted_page_url = JSON.parse(response.body)['live_stream']['hosted_page_url']
        ap "There it is! If you like to see a hosted page and a player with your Live Stream, feel free to open this URL in your browser:", color: {string: :yellowish}
        puts "\n#{hosted_page_url}"
        puts "(Note: In the sandbox environment, this URL and other player-related data are simulated.)\n\n" if @settings["api_base_url"].match(/\-sandbox/)
        ap "And by the way: Now would also be a good time to tweet it or announce it somewhere programmatically.", color: {string: :yellowish}
        next_step += 1
      else
        puts "\n"
        ap "Unfortunately something went wrong provisioning the Player of your Live Stream. Let's continue with the next step.", color: {string: :red}
      end

      # Next Step: stop
      puts "\n"
      ap "Step #{next_step}: Ok, let's go ahead and stop the Live Stream again.", color: {string: :purpleish}
      puts "\n"
      @cli.ask next_step_text
      stop id
      next_step += 1

      # Next Step: delete
      puts "\n"
      ap "Step #{next_step}: And finally we clean up a bit and delete the Live Stream.", color: {string: :purpleish}
      puts "\n"
      @cli.ask next_step_text
      delete id
      next_step += 1

      puts "\n"
      ap "Workflow finished!", color: {string: :yellowish}
      puts "\n"
      @cli.ask "Thank you! Hit Enter to return to Main Menu!"
    # Error handling if the Live Stream failed to start
    else
      puts "\n"
      ap "Unfortunately something went wrong starting the Live Stream. Going to delete the Live Stream and return to Main Menu!", color: {string: :red}
      puts "\n"
      delete id
      menu
      return
    end
  end

  # calling the WSC API, using the Ruby Net::HTTP library
  #
  # available options (options[:NAME]):
  #
  # method: the HTTP method you want to use to call the API
  #  supported values: :get, :post, :patch, :put, :delete
  #
  # endpoint: the endpoint to call, default: live_streams (optional)
  #  supported values: any existing endpoint of the Wowza Streaming Cloud API
  #
  # id: the unique id of a live stream (optional)
  #  supported values: any existing id of a live stream of your account on https://cloud.wowza.com
  #
  # action: the action you want to call (optional)
  #  supported values: '/state', '/start', '/reset', '/stop', '/thumbnail_url'
  #
  # body: a JSON interpretation of live stream settings you want to push to the API (optional)
  #  supported values: any JSON that contains valid properties and is wrapped in a root element 'live_streams'
  #
  def call_api options = {}
    endpoint = options[:endpoint] ||'live_streams'
    api_uri = URI("#{@settings["api_base_url"]}/api/#{@settings["api_version"]}/#{endpoint}/#{options[:id]}#{options[:action]}")
    response = nil
    ShellSpinner "calling #{options[:method].upcase} #{api_uri.to_s}" do # each call to the external API is showing a little spinner while running
      puts "\n"
      Net::HTTP.start(api_uri.host, api_uri.port, use_ssl: api_uri.scheme == 'https') do |http|
        case options[:method]
        when :get
          request = Net::HTTP::Get.new api_uri.path
        when :post
          request = Net::HTTP::Post.new api_uri.path
          request.body = options[:body].to_json
        when :patch
          request = Net::HTTP::Patch.new api_uri.path
          request.body = options[:body].to_json
        when :put
          request = Net::HTTP::Put.new api_uri.path
          request.body = options[:body].to_json
        when :delete
          request = Net::HTTP::Delete.new api_uri.path
        end
        # add keys to Header for authentication
        request['wsc-api-key'] = @keys["api_key"]
        request['wsc-access-key'] = @keys["api_access_key"]
        # add the right Content-Type to Header (always required when pushing data)
        request["Content-Type"] = "application/json"

        response = http.request(request)
      end
    end

    # print response code and headers if debug setting turned on
    if @settings["debug"]
      puts "###### DEBUG #####"
      puts "HTTP Response Code: #{response.code}\n"
      puts "HTTP Response Headers:"
      response.each_header do |key, value|
        p "#{key} => #{value}"
      end
      puts "###### DEBUG #####\n"
    end

    return response
  end


  ##############################################################################
  # HELPER METHODS
  ##############################################################################

  # loads configuration from /config folder
  # error handling for missing properties
  #
  def init_config
    # load API keys YAML from config
    @keys = YAML.load_file('config/keys.yml')

    # error handling for missing API keys
    if @keys["api_key"].nil? || @keys["api_key"].empty?
      puts "No api_key specified in config/keys.yml!\n\n"
      quit
    end
    if @keys["api_access_key"].nil? || @keys["api_access_key"].empty?
      puts "No api_access_key specified in config/keys.yml!\n\n"
      quit
    end

    # load settings YAML from config
    @settings = YAML.load_file('config/settings.yml')

    # error handling for missing app settings
    if @settings["api_base_url"].nil? || @settings["api_base_url"].empty?
      puts "No api_base_url specified in config/settings.yml!\n\n"
      quit
    end
    if @settings["api_version"].nil? || @settings["api_version"].empty?
      puts "No api_version specified in config/settings.yml!\n\n"
      quit
    end
  end

  # launches application:
  # showing Wowza Intro and opening Main Menu
  #
  def launch!
    intro
    menu
  end

  # prints Wowza Intro
  #
  def intro
    puts "\n"
    puts "--------------------------------------------------------------------------------------------------------------\n\n"
    puts "              Wowza Streaming Cloud API - An example application for the Live Stream workflow  \n\n"
    puts "--------------------------------------------------------------------------------------------------------------\n\n"
    file = File.open("assets/wsc-logo-ascii.txt", "r")
    puts "#{file.read}\n"
    puts "--------------------------------------------------------------------------------------------------------------\n\n"
    puts "                                                   © 2013-#{Time.now.year} Wowza Media Systems™, LLC. All rights reserved.\n\n"
    if @settings["api_base_url"].match(/\-sandbox/)
      puts "\n"
      ap "Note: You are using the sandbox environment.", color: {string: :yellowish}
      puts "\n"
      puts "The API sandbox allows you to test and develop streaming workflows in a safe, separate environment that doesn't affect your Wowza Streaming Cloud production environment."
      puts "The sandbox is fully functional and can do everything that the real API does, except that it creates simulated live streams (players and hosted pages) and transcoders, not real ones."
      puts "Because the sandbox doesn't use real resources, you don't accrue charges when you use it."
      puts "\n"
    end
  end

  # displays Main Menu
  #
  def menu
    options = {
      count:         "Show the number of live streams in your account",
      list:          "List all live streams of your account",
      create:        "Create a live stream with pre-configured settings         => data/live_stream/encoder_types/*.json",
      show:          "Show the details of an existing live stream",
      update:        "Update a live stream with pre-configured settings         => data/live_stream/update_example.json",
      start:         "Start a live stream                                       => only for Live Streams with the state 'stopped'",
      reset:         "Reset a live stream                                       => only for Live Streams with the state 'started'",
      stop:          "Stop a live stream                                        => only for Live Streams with the state 'started'",
      state:         "Show the current state of a live stream",
      thumbnail_url: "Show the thumbnail URL of a live stream                  => only for Live Streams with the state 'started'",
      delete:        "Delete a live stream",
      workflow:      "Run the pre-configured live stream workflow",
      quit:          "Quit :("
    }
    @cli = HighLine.new
    loop do
      @cli.choose do |menu|
        puts "\n"
        ap "Main Menu"
        puts "\n"
        menu.prompt = "\nEnter the number for the action you want to execute."
        options.each_with_index do |(method, description), index|
          menu.choice(description) do
            puts "\n"
            puts "-----------------------------"
            ap "Processing Option #{index+1}: #{description}", color: {string: :cyanish}
            puts "-----------------------------\n\n"
            # list action with submenu for simple or detailed
            case method
            when :list
              @cli.choose do |menu|
                ap "Submenu"
                puts "\n"
                menu.prompt = "\nSimple or detailed? Please enter a number!"
                menu.choice("Simple (just id and name)") do
                  list simple: true
                end
                menu.choice("Detailed (full JSON response)") do
                  list
                end
                menu.choice("Back to Main Menu") do
                  menu
                end
              end
            # choose one of the pre-configured live stream setups
            when :create
              # build submenu with files from the data/live_streams directory
              @cli.choose do |menu|
                puts "\n"
                ap "Submenu"
                puts "\n"
                menu.prompt = "\nWhat pre-configured camera or encoder will you use to connect to Wowza Streaming Cloud? Please enter a number!"
                # let's search through the example directory
                Dir.glob("./data/live_stream/encoder_types/*.json").each do |f|
                  encoder_json = JSON.parse(File.read(f))
                  name = encoder_json['live_stream']['name'] rescue 'Unknown Name. Please change the content of the JSON and add a name!'
                  menu.choice(name) do
                    create encoder_json
                  end
                end
                menu.choice("Back to Main Menu") do
                  menu
                end
              end
            # actions without additional parameters
            when :workflow, :count, :quit
              send method
            # actions that needs a Live Stream ID to be selected from a list
            else
              # build submenu with existing Live Streams / call list live streams
              response = call_api method: :get
              json = JSON.parse(response.body)
              @cli.choose do |menu|
                puts "\n"
                ap "Submenu"
                puts "\n"
                menu.prompt = "\nWhich Live Stream do you want to call? Please enter a number!"
                json['live_streams'].each do |live_stream|
                  menu.choice("#{live_stream['name']} (#{live_stream['id']})") do
                    send method, live_stream['id']
                  end
                end
                menu.choice("Back to Main Menu") do
                  menu
                end
              end
            end
            puts "\n-----------------------------\n"
          end
        end
      end
    end
  end

  # exits the application
  #
  def quit
    puts "Goodbye!\n"
    exit
  end

end

# init and launch application
#
lswe = LiveStreamApiExample.new
lswe.launch!
